cpp/src/buffer/buffer.cpp

```diff
@@ -98,19 +74,18 @@ void buffer_copy(
         return;  // Nothing to copy.
     }
 
+    // We have to sync both before *and* after the memcpy. Otherwise, `src.stream()`
+    // might deallocate `src` before the memcpy enqueued on `dst.stream()` has completed.
     cuda_stream_join(std::array{dst.stream()}, std::array{src.stream()});
-    RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
-        dst.data() + dst_offset,
-        src.data() + src_offset,
-        size,
-        cudaMemcpyDefault,
-        dst.stream()
-    ));
-
-    // Override the event to track the async copy.
-    if (attach_cuda_event) {
-        src.override_event(CudaEvent::make_shared_record(src.stream()));
-        dst.override_event(CudaEvent::make_shared_record(dst.stream()));
-    }
+    dst.write_access(dst.stream(), [&](std::byte* dst_data) {
+        RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
+            dst_data + dst_offset,
+            src.data() + src_offset,
+            size,
+            cudaMemcpyDefault,
+            dst.stream()
+        ));
+    });
+    cuda_stream_join(std::array{src.stream()}, std::array{dst.stream()});
```

Comment:
We could potentially keep a (per-context and) per-thread cache of events around to use inside `cuda_stream_join`. I think because we never query the context of the stream, and are just using the runtime api, it only needs to be per-thread.

Reply:
But don't we get to the whole _CUDA static object problem_?

Reply:
A few options:

1. Make them per-`BufferResource` object which is not static
2. Make them static but just don't destruct them

But I suspect this is a microopt we should revisit if we see things in the profiles.